model/src/main/scala/explore/data/KeyedIndexedList.scala

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package explore.data
+
+import scala.collection.immutable.ListMap
+
+import cats.implicits._
+import cats.kernel.Eq
+import monocle.Lens
+
+// Each element has a unique Key.
+// Elements can be accessed by Key in O(1).
+// Position in the list (index) can be accessed by Key in O(1).
+case class KeyedIndexedList[K, A] private (private val list: ListMap[K, (A, Int)]) {
+  def getElemAndIndex(id: K): Option[(A, Int)] = list.get(id)
+  def getElement(id:      K): Option[A]        = list.get(id).map(_._1)
+  def getIndex(id:        K): Option[Int]      = list.get(id).map(_._2)
+
+  def elements: Iterable[A] = list.values.map(_._1)
+
+  def toList: List[A] = elements.toList
+
+  def length: Int = list.size
+
+  def removed(key: K): KeyedIndexedList[K, A] =
+    getIndex(key)
+      .fold(this)(idx =>
+        KeyedIndexedList.unsafeFromListMap(
+          list
+            .removed(key)
+            .map(_ match {
+              case (id, (a, i)) if i < idx => (id, (a, i))
+              case (id, (a, i)) if i > idx => (id, (a, i - 1))
+            })
+        )
+      )
+
+  def contains(key: K): Boolean = list.contains(key)
+
+  def exists(p: A => Boolean): Boolean = elements.exists(p)
+
+  def take(n: Int): KeyedIndexedList[K, A] = KeyedIndexedList.unsafeFromListMap(list.take(n))
```

Comment:
Should these be unsafe too?

Reply:
I don't think so. The unsafe constructor is unsafe because there's no guarantee that the indices are correct. But we can keep track of that here.

That being said, the `drop` method is wrong: indices should be adjusted.